src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -0,0 +1,1187 @@
+ * issues.
+ *
+ * @since 19
+ */
+/*non-public*/
+public final class Carrier {
+    /**
+     * Class file version.
+     */
+    private static final int CLASSFILE_VERSION = VM.classFileVersion();
+
+    /**
+     * Lookup used to define and reference the carrier object classes.
+     */
+    private static final Lookup LOOKUP;
+
+    /**
+     * Maximum number of components in a carrier (based on the maximum
+     * number of args to a constructor.)
+     */
+    private static final int MAX_COMPONENTS = 255 - /* this */ 1;
+
+    /**
+     * Maximum number of components in a CarrierClass.
+     */
+    private static final int MAX_OBJECT_COMPONENTS = 32;
+
+    /**
+     * Stable annotation.
+     */
+    private static final String STABLE = "jdk/internal/vm/annotation/Stable";
+    private static final String STABLE_SIG = "L" + STABLE + ";";
+
+    /**
+     * Number of integer slots used by a long.
+     */
+    private static final int LONG_SLOTS = 2;
+
+    /*
+     * Initialize {@link MethodHandle} constants.
+     */
+    static {
+        LOOKUP = MethodHandles.lookup();
+
+        try {
+            FLOAT_TO_INT = LOOKUP.findStatic(Float.class, "floatToRawIntBits",
+                    methodType(int.class, float.class));
+            INT_TO_FLOAT = LOOKUP.findStatic(Float.class, "intBitsToFloat",
+                    methodType(float.class, int.class));
+            DOUBLE_TO_LONG = LOOKUP.findStatic(Double.class, "doubleToRawLongBits",
+                    methodType(long.class, double.class));
+            LONG_TO_DOUBLE = LOOKUP.findStatic(Double.class, "longBitsToDouble",
+                    methodType(double.class, long.class));
+        } catch (ReflectiveOperationException ex) {
+            throw new AssertionError("carrier static init fail", ex);
+        }
+    }
+
+    /*
+     * float/double conversions.
+     */
+    private static final MethodHandle FLOAT_TO_INT;
+    private static final MethodHandle INT_TO_FLOAT;
+    private static final MethodHandle DOUBLE_TO_LONG;
+    private static final MethodHandle LONG_TO_DOUBLE;
+
+    /**
+     * long signature descriptor.
+     */
+    private static final String LONG_DESCRIPTOR =
+            Type.getDescriptor(long.class);
+
+    /**
+     * int signature descriptor.
+     */
+    private static final String INT_DESCRIPTOR =
+            Type.getDescriptor(int.class);
+
+    /**
+     * Object signature descriptor.
+     */
+    private static final String OBJECT_DESCRIPTOR =
+            Type.getDescriptor(Object.class);
+
+    /**
+     * Given a constructor {@link MethodHandle} recast and reorder arguments to
+     * match shape.
+     *
+     * @param carrierShape  carrier shape
+     * @param constructor   carrier constructor to reshape
+     *
+     * @return constructor with arguments recasted and reordered
+     */
+    private static MethodHandle reshapeConstructor(CarrierShape carrierShape,
```

Comment:
I like how reshape constructors and components is now shared across both implementations

Reply:
Refactoring is just another tool for the lazy. ;-)